### PR TITLE
Enable bucket pruning for IN predicates

### PR DESCRIPTION
**Pull request by gparai**
Given table `t` bucketed by column `x` and a query `SELECT * FROM t WHERE x = 5`, Presto compute bucket ID for x = 5 and reduces table scan to just that bucket. This technique is called bucket pruning.

This change extends bucket pruning to predicates where multiple values match. For queries with predicates like `x IN (5, 7)` and `x = 5 OR x = 7`, Presto will compute bucket IDs for values 5 and 7 and reduce table scan to just these buckets. 

This logic doesn't apply to predicates like `x BETWEEN 5 AND 7` or `x > 5 AND x < 10`.

CC: @prestodb/aria 

== NO RELEASE NOTE ==

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 1]
```diff
@@ -300,6 +301,15 @@ private static int hashBytes(int initialValue, Slice bytes)
             return Optional.of(new HiveBucketFilter(ImmutableSet.of(singleBucket.getAsInt())));
         }
 
+        Optional<Map<ColumnHandle, Set<NullableValue>>> columnValueSetBindings = TupleDomain.extractFixedValuesFromRanges(effectivePredicate);
+        if (!columnValueSetBindings.isPresent()) {
+            return Optional.empty();
+        }
+        Set<Integer> buckets = getHiveBuckets(table, columnValueSetBindings.get());
+        if (buckets.size() > 0) {
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: @gparai What if there is no match? The desired behavoir would be to not read any file. Is this how it would work?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 2]
```diff
@@ -300,6 +301,15 @@ private static int hashBytes(int initialValue, Slice bytes)
             return Optional.of(new HiveBucketFilter(ImmutableSet.of(singleBucket.getAsInt())));
         }
 
+        Optional<Map<ColumnHandle, Set<NullableValue>>> columnValueSetBindings = TupleDomain.extractFixedValuesFromRanges(effectivePredicate);
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: @gparai It seems like the logic in lines 295-302 would be covered by a new more generic logic, no? E.g. that logic processing the case where the filter matches exactly one bucket, but the new logic allows for multiple buckets. Any reason to keep both?

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java [hunk 3]
```diff
@@ -119,6 +120,27 @@ private TupleDomain(Optional<Map<T, Domain>> domains)
...
+     * Returns an empty Optional if the Domain is none.
+     */
+    public static <T> Optional<Map<T, Set<NullableValue>>> extractFixedValuesFromRanges(TupleDomain<T> tupleDomain)
+    {
+        if (!tupleDomain.getDomains().isPresent()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(tupleDomain.getDomains().get().entrySet().stream()
+                .filter(entry -> entry.getValue().getValues().getType().isOrderable())
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: @gparai this is a bit difficult to read and it seems like a lot of computation is repeated unnecessarily; consider rewriting for clarity and efficiency:

```
    public static <T> Optional<Map<T, Set<NullableValue>>> extractFixedValuesFromRanges(TupleDomain<T> tupleDomain)
    {
        if (!tupleDomain.getDomains().isPresent()) {
            return Optional.empty();
        }

        Map<T, Set<NullableValue>> fixedValues = new HashMap<>();
        for (Map.Entry<T, Domain> entry : tupleDomain.getDomains().get().entrySet()) {
            Optional<Set<NullableValue>> values = extractFixedValues(entry.getValue());
            if (values.isPresent()) {
                fixedValues.put(entry.getKey(), values.get());
            }
        }

        return Optional.of(fixedValues);
    }

    private static Optional<Set<NullableValue>> extractFixedValues(Domain domain)
    {
        ValueSet values = domain.getValues();
        if (!values.getType().isOrderable()) {
            return Optional.empty();
        }

        List<Range> ranges = values.getRanges().getOrderedRanges();
        boolean allSingleValue = ranges.stream().allMatch(Range::isSingleValue);
        if (!allSingleValue) {
            return Optional.empty();
        }

        return Optional.of(ranges.stream()
                .map(range -> new NullableValue(domain.getType(), range.getSingleValue()))
                .collect(Collectors.toSet()));
    }
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 4]
```diff
@@ -321,7 +334,102 @@ private static int hashBytes(int initialValue, Slice bytes)
...
+        }
+
+        // Verify the bucket column types are supported
+        for (String column : bucketColumns) {
+            if (!SUPPORTED_TYPES_FOR_BUCKET_FILTER.contains(hiveTypes.get(column))) {
+                return Optional.empty();
+            }
+        }
+
+        List<Pair<String, TypeInfo>> columnTypeInfoPairs = new ArrayList<>();
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: this logic can be simplified using streaming APIs:

```
    private static Optional<List<Pair<String, TypeInfo>>> getBucketColumnTypeInfos(Table table)
    {
        List<String> bucketColumns = table.getStorage().getBucketProperty().get().getBucketedBy();

        // Verify the bucket column types are supported
        Map<String, HiveType> hiveTypes = table.getDataColumns().stream()
                .collect(toImmutableMap(Column::getName, Column::getType));
        for (String column : bucketColumns) {
            if (!SUPPORTED_TYPES_FOR_BUCKET_FILTER.contains(hiveTypes.get(column))) {
                return Optional.empty();
            }
        }

        return Optional.of(bucketColumns.stream()
                .map(column -> new ImmutablePair<>(column, hiveTypes.get(column).getTypeInfo()))
                .collect(toImmutableList()));
    }
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 5]
```diff
@@ -291,13 +295,22 @@ private static int hashBytes(int initialValue, Slice bytes)
...
+        /*Optional<Map<ColumnHandle, NullableValue>> bindings = TupleDomain.extractFixedValues(effectivePredicate);
         if (!bindings.isPresent()) {
             return Optional.empty();
         }
         OptionalInt singleBucket = getHiveBucket(table, bindings.get());
         if (singleBucket.isPresent()) {
             return Optional.of(new HiveBucketFilter(ImmutableSet.of(singleBucket.getAsInt())));
+        }*/
+
+        Optional<Map<ColumnHandle, Set<NullableValue>>> columnValueSetBindings = TupleDomain.extractFixedValuesFromRanges(effectivePredicate);
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: columnValueSetBindings -> bindings for brevity

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 6]
```diff
@@ -321,7 +334,102 @@ private static int hashBytes(int initialValue, Slice bytes)
...
+        if (bindings.isEmpty()) {
+            return OptionalInt.empty();
+        }
+
+        // Check that we have bindings for all bucket columns
+        if (bindings.size() != bucketColumnTypeInfos.size()) {
+            return OptionalInt.empty();
+        }
+
+        ImmutableList.Builder<TypeInfo> typeInfos = ImmutableList.builder();
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: The list of types is the same for each set of values. No need to re-compute it again and again. Consider refactoring the logic to (1) get a list of types; (2) check if there are bindings for all necessary columns; (3) generate all possible permutations of bucketed_by column values and compute buckets for each.

```
    private static Optional<Set<Integer>> getHiveBuckets(Table table, Map<ColumnHandle, Set<NullableValue>> bindings)
    {
        Optional<List<Pair<String, TypeInfo>>> nameToType = getBucketColumnTypeInfos(table);

        if (!nameToType.isPresent()) {
            return Optional.empty();
        }

        Map<String, Set<NullableValue>> nameToBindings = bindings.entrySet().stream()
                .collect(toImmutableMap(entry -> ((HiveColumnHandle) entry.getKey()).getName(), Entry::getValue));

        ImmutableList.Builder<Set<NullableValue>> orderedBindingsBuilder = ImmutableList.builder();
        for (Pair<String, TypeInfo> typeInfoPair : nameToType.get()) {
            String columnName = typeInfoPair.getKey();
            if (!nameToBindings.containsKey(columnName)) {
                return Optional.empty();
            }
            orderedBindingsBuilder.add(nameToBindings.get(columnName));
        }

        List<Set<NullableValue>> orderedBindings = orderedBindingsBuilder.build();

        int bucketCount = table.getStorage().getBucketProperty().get().getBucketCount();
        List<TypeInfo> types = nameToType.get().stream()
                .map(Pair::getRight)
                .collect(toImmutableList());

        ImmutableSet.Builder<Integer> buckets = ImmutableSet.builder();
        getHiveBuckets(new Object[types.size()], 0, orderedBindings, bucketCount, types, buckets);

        return Optional.of(buckets.build());
    }

    private static void getHiveBuckets(
            Object[] values,
            int valuesCount,
            List<Set<NullableValue>> bindings,
            int bucketCount,
            List<TypeInfo> typeInfos,
            ImmutableSet.Builder<Integer> buckets)
    {
        if (valuesCount == typeInfos.size()) {
            buckets.add(getHiveBucket(bucketCount, typeInfos, values));
            return;
        }

        for (NullableValue value : bindings.get(valuesCount)) {
            values[valuesCount] = value.getValue();
            getHiveBuckets(values, valuesCount + 1, bindings, bucketCount, typeInfos, buckets);
        }
    }
```

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 7]
```diff
@@ -731,6 +731,38 @@ public void testPushdownSubfieldsAndJoinReordering()
...
+    {
+        QueryRunner queryRunner = getQueryRunner();
+        try {
+            queryRunner.execute("CREATE TABLE test_bucket_filter1 WITH (bucket_count = 11, bucketed_by = ARRAY['key1']) AS " +
+                    "SELECT orderkey AS key1, comment AS value1 FROM orders");
+            queryRunner.execute("CREATE TABLE test_bucket_filter2 WITH " +
+                    "(bucket_count = 13, partitioned_by = ARRAY[ 'orderstatus' ], bucketed_by = ARRAY[ 'key1', 'key2' ]) AS " +
+                    "SELECT orderkey AS key1, custkey AS key2, comment AS value1, orderstatus FROM orders");
+
+            // Remaining buckets after pruning - 2
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: I made the following changes to get these tests to pass:

```
            // Remaining buckets after pruning - 2
            assertPlan(getSession(),
                    "SELECT value1 FROM test_bucket_filter1 WHERE key1 = 101 OR key1 = 60000",
                    anyTree(strictTableScan("test_bucket_filter1", identityMap("key1", "value1"))),
                    plan -> assertTableBucketingLayout(plan, "test_bucket_filter1", 11, ImmutableSet.of(2, 6)));

            // Remaining buckets after pruning - 2
            assertPlan(getSession(),
                    "SELECT value1 FROM test_bucket_filter2 WHERE key1 IN (101, 60000) AND key2 IN (1426, 280)",
                    anyTree(strictTableScan("test_bucket_filter2", identityMap("key1", "key2", "value1"))),
                    plan -> assertTableBucketingLayout(plan, "test_bucket_filter2", 13, ImmutableSet.of(8, 6, 7, 5)));


    private void assertTableBucketingLayout(Plan plan, String tableName, int readBucketCount, Set<Integer> bucketsToKeep)
    {
        TableScanNode tableScan = searchFrom(plan.getRoot())
                .where(node -> isTableScanNode(node, tableName))
                .findOnlyElement();

        assertTrue(tableScan.getTable().getLayout().isPresent());
        HiveTableLayoutHandle layoutHandle = (HiveTableLayoutHandle) tableScan.getTable().getLayout().get();

        assertEquals(layoutHandle.getBucketHandle().get().getReadBucketCount(), readBucketCount);
        assertEquals(layoutHandle.getBucketFilter().get().getBucketsToKeep(), bucketsToKeep);
    }
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 1 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: We would always get some bucket for the given binding unless something was not right/valid in which case it would be `Optional.empty()`. I have removed this check and rely on the optional instead.

Comment (mbasmanova) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: @gparai You are right. Thanks for clarifying.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 2 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: The new logic applies to SortedRangeSet. I was not sure if for the single bucket case we could end up with predicates having EquatableValueSet/AllOrNoneValueSet hence I had left it alone. For the single bucket case, it looks like we generate the SortedRangeSet so keeping only the new logic should be fine.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java (hunk 3 above)
Comment (gparai) [reply presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: Done

File: presto-main/src/main/java/com/facebook/presto/server/PluginManager.java [hunk 8]
```diff
@@ -150,6 +150,9 @@ public void loadPlugins()
         }
 
         for (String plugin : plugins) {
+            if (plugin.startsWith("fbraptor") || plugin.startsWith("shardedmysql")) {
```
[the remark this answers is not in the document]
Comment (gparai) [reply presto-main/src/main/java/com/facebook/presto/server/PluginManager.java]: This was a temp commit with some unrelated changes. Removed it.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 9]
```diff
@@ -291,13 +295,22 @@ private static int hashBytes(int initialValue, Slice bytes)
             return Optional.empty();
         }
 
-        Optional<Map<ColumnHandle, NullableValue>> bindings = TupleDomain.extractFixedValues(effectivePredicate);
+        /*Optional<Map<ColumnHandle, NullableValue>> bindings = TupleDomain.extractFixedValues(effectivePredicate);
```
[the remark this answers is not in the document]
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Done

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 10]
```diff
@@ -321,7 +334,102 @@ private static int hashBytes(int initialValue, Slice bytes)
...
+        }
+
+        List<Pair<String, TypeInfo>> columnTypeInfoPairs = new ArrayList<>();
+        for (int i = 0; i < bucketColumns.size(); i++) {
+            String column = bucketColumns.get(i);
+            columnTypeInfoPairs.add(new ImmutablePair<>(column, hiveTypes.get(column).getTypeInfo()));
+        }
+        return Optional.of(columnTypeInfoPairs);
+    }
+    /*private static OptionalInt getHiveBucket(Table table, Map<ColumnHandle, NullableValue> bindings)
```
[the remark this answers is not in the document]
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Done

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 4 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Thanks! Done

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 5 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Done

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 6 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Thanks! Done.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 7 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: It works! Thanks! I changed the layout to display bucketsToKeep which was not displayed earlier.

**Comment by gparai** [comment]
@mbasmanova I have addressed your review comments. Please take a look and let me know your thoughts.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 11]
```diff
@@ -321,49 +323,78 @@ private static int hashBytes(int initialValue, Slice bytes)
         return Optional.of(new HiveBucketFilter(builder.build()));
     }
 
-    private static OptionalInt getHiveBucket(Table table, Map<ColumnHandle, NullableValue> bindings)
+    private static Optional<Set<Integer>> getHiveBuckets(Table table, Map<ColumnHandle, Set<NullableValue>> bindings)
     {
-        if (bindings.isEmpty()) {
-            return OptionalInt.empty();
-        }
+        Optional<List<SimpleImmutableEntry<String, TypeInfo>>> nameToType = getBucketColumnTypeInfos(table);
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: we tend not to use Pair-like objects; here, I suggest inlining getBucketColumnTypeInfos like so:

```
        List<String> bucketColumns = table.getStorage().getBucketProperty().get().getBucketedBy();
        if (bucketColumns.isEmpty()) {
            return Optional.empty();
        }

        Map<String, HiveType> hiveTypes = table.getDataColumns().stream()
                .collect(toImmutableMap(Column::getName, Column::getType));

        // Verify the bucket column types are supported
        for (String column : bucketColumns) {
            if (!SUPPORTED_TYPES_FOR_BUCKET_FILTER.contains(hiveTypes.get(column))) {
                return Optional.empty();
            }
        }

        Map<String, Set<NullableValue>> nameToBindings = bindings.entrySet().stream()
                .collect(toImmutableMap(entry -> ((HiveColumnHandle) entry.getKey()).getName(), Entry::getValue));

        ImmutableList.Builder<Set<NullableValue>> orderedBindingsBuilder = ImmutableList.builder();
        for (String columnName : bucketColumns) {
            if (!nameToBindings.containsKey(columnName)) {
                return Optional.empty();
            }
            orderedBindingsBuilder.add(nameToBindings.get(columnName));
        }

        List<Set<NullableValue>> orderedBindings = orderedBindingsBuilder.build();

        int bucketCount = table.getStorage().getBucketProperty().get().getBucketCount();
        List<TypeInfo> types = bucketColumns.stream()
                .map(hiveTypes::get)
                .map(HiveType::getTypeInfo)
                .collect(toImmutableList());

        ImmutableSet.Builder<Integer> buckets = ImmutableSet.builder();
        getHiveBuckets(new Object[types.size()], 0, orderedBindings, bucketCount, types, buckets);

        return Optional.of(buckets.build());
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 12]
```diff
@@ -1937,7 +1938,7 @@ public ConnectorPushdownFilterResult pushdownFilter(ConnectorSession session, Co
                                 hivePartitionResult.getBucketHandle(),
                                 hivePartitionResult.getBucketFilter(),
                                 true,
-                                createTableLayoutString(session, tableName, hivePartitionResult.getBucketHandle(), decomposedFilter.getRemainingExpression(), domainPredicate))),
+                                createTableLayoutString(session, tableName, hivePartitionResult.getBucketHandle(), hivePartitionResult.getBucketFilter(), decomposedFilter.getRemainingExpression(), domainPredicate))),
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: would you put this change into a separate commit?

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java [hunk 13]
```diff
@@ -119,6 +120,46 @@ private TupleDomain(Optional<Map<T, Domain>> domains)
                 .collect(toLinkedMap(Map.Entry::getKey, entry -> new ConstantExpression(entry.getValue().getNullableSingleValue(), entry.getValue().getType()))));
     }
 
+    /**
+     * Extract all column constraints that require a set of one (or more) values in their respective Domains. All the
+     * ranges (see {@link SortedRangeSet}) must have fixed values.
+     * Returns an empty Optional if the Domain is none.
+     */
+    public static <T> Optional<Map<T, Set<NullableValue>>> extractFixedValuesFromRanges(TupleDomain<T> tupleDomain)
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: extractFixedValuesFromRanges  -> extractFixedValueSets for clarity; 'from ranges' doesn't differentiate this method for others because they all extract information from rages;

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java [hunk 14]
```diff
@@ -119,6 +120,46 @@ private TupleDomain(Optional<Map<T, Domain>> domains)
...
+            Optional<Set<NullableValue>> values = extractFixedValuesFromOrderableDomain(entry.getValue());
+            if (values.isPresent()) {
+                fixedValues.put(entry.getKey(), values.get());
+            }
+        }
+
+        return Optional.of(fixedValues);
+    }
+
+    private static Optional<Set<NullableValue>> extractFixedValuesFromOrderableDomain(Domain domain)
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: extractFixedValuesFromOrderableDomain -> extractFixedValueSet (singular set here)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 15]
```diff
@@ -731,6 +731,35 @@ public void testPushdownSubfieldsAndJoinReordering()
         }
     }
 
+    @Test
+    public void testBucketFiltering()
+    {
+        QueryRunner queryRunner = getQueryRunner();
+        try {
+            queryRunner.execute("CREATE TABLE test_bucket_filter1 WITH (bucket_count = 11, bucketed_by = ARRAY['key1']) AS " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: - Let's try to come up with more descriptive names for readability. For example let's create orders_bucketed table first bucketed on orderkey, then on orderkey and custkey.
- Let's test both positive and negative scenarios.
- CREATE TABLE statement goes before `try` block. This way `finally` block can use DROP TABLE without IF EXISTS as the table will have to exist if `finally` is reached.

```
QueryRunner queryRunner = getQueryRunner();
        queryRunner.execute("CREATE TABLE orders_bucketed WITH (bucket_count = 11, bucketed_by = ARRAY['orderkey']) AS " +
                "SELECT * FROM orders");

        try {
            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 100",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 100 OR orderkey = 101",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1, 2)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey IN (100, 101, 133)",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1, 2)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey > 100",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey != 100",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));
        }
        finally {
            queryRunner.execute("DROP TABLE orders_bucketed");
        }

        queryRunner.execute("CREATE TABLE orders_bucketed WITH (bucket_count = 11, bucketed_by = ARRAY['orderkey', 'custkey']) AS " +
                "SELECT * FROM orders");

        try {
            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 101 AND custkey = 280",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey IN (101, 71) AND custkey = 280",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1, 6)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey IN (101, 71) AND custkey IN (280, 34)",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1, 2, 6, 8)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 101 AND custkey = 280 AND orderstatus <> '0'",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertBucketFilter(plan, "orders_bucketed", 11, ImmutableSet.of(1)));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 101",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE custkey = 280",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));

            assertPlan(getSession(), "SELECT * FROM orders_bucketed WHERE orderkey = 101 AND custkey > 280",
                    anyTree(PlanMatchPattern.tableScan("orders_bucketed")),
                    plan -> assertNoBucketFilter(plan, "orders_bucketed", 11));
        }
        finally {
            queryRunner.execute("DROP TABLE orders_bucketed");
        }
```

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 16]
```diff
@@ -1950,6 +1950,43 @@ private void testBucketedExecution(Session session)
                 "SELECT custkey, COUNT(*) FROM orders GROUP BY custkey");
     }
 
+    @Test
+    public void testBucketedFilterExecution()
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: testBucketedFilterExecution  -> testBucketPruning

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 17]
```diff
@@ -731,6 +731,35 @@ public void testPushdownSubfieldsAndJoinReordering()
         }
     }
 
+    @Test
+    public void testBucketFiltering()
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: testBucketFiltering  -> testBucketPruning

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 18]
```diff
@@ -1950,6 +1950,43 @@ private void testBucketedExecution(Session session)
...
+    {
+        testBucketedFilterExecution(getSession());
+    }
+
+    private void testBucketedFilterExecution(Session session)
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: inline this method

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 19]
```diff
@@ -1950,6 +1950,43 @@ private void testBucketedExecution(Session session)
...
+    {
+        try {
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: - let's make this test run the same queries as the other test

```
        getQueryRunner().execute("CREATE TABLE orders_bucketed WITH (bucket_count = 11, bucketed_by = ARRAY['orderkey']) AS " +
                "SELECT * FROM orders");

        try {
            assertQuery("SELECT * FROM orders_bucketed WHERE orderkey = 100", "SELECT * FROM orders WHER orderkey = 100");
            assertQuery("SELECT * FROM orders_bucketed WHERE orderkey = 100 OR orderkey = 101", "SELECT * FROM orders WHERE orderkey = 100 OR orderkey = 101");
            // TODO add more tests
        }
        finally {
            assertUpdate("DROP TABLE orders_bucketed");
        }
```

**Comment by mbasmanova** [comment]
CC: @wenleix

**Comment by mbasmanova** [comment]
@gparai Re: commit message, I suggest to shorten it to `Enable bucket pruning for IN predicates` or expand to provide more details.

```
Enable bucket pruning for IN predicates

Given table t bucketed by column x and a query SELECT * FROM t WHERE x = 5, Presto computes 
bucket ID for x = 5 and reduces table scan to just that bucket. This technique is called bucket 
pruning.

This change extends bucket pruning to predicates where multiple values match. For queries with 
predicates like x IN (5, 7) and x = 5 OR x = 7, Presto will compute bucket IDs for values 5 and 7 
and reduce table scan to just these buckets.

This functionality doesn't apply to predicates like x BETWEEN 5 AND 7 or x > 5 AND x < 10.
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 11 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Sure.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 12 above)
Comment (gparai) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Done

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java (hunk 13 above)
Comment (gparai) [reply presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: Done

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java (hunk 14 above)
Comment (gparai) [reply presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 15 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Thanks for reminder about negative test scenarios!

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 16 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 17 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 19 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Done

**Comment by gparai** [comment]
@mbasmanova I have addressed your review comments. Please take a look and let me know your thoughts.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 20]
```diff
@@ -1950,6 +1951,54 @@ private void testBucketedExecution(Session session)
...
+    @Test
+    public void testBucketPruning()
+    {
+        Session session = getSession();
+        QueryRunner queryRunner = getQueryRunner();
+        queryRunner.execute("CREATE TABLE orders_bucketed WITH (bucket_count = 11, bucketed_by = ARRAY['orderkey']) AS " +
+                "SELECT * FROM orders");
+
+        try {
+            assertQuery(session, "SELECT count(*) FROM orders_bucketed WHERE orderkey = 100", "SELECT 1");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: @gparai It would be better to use SELECT * queries. count(*) only confirms that row counts are correct, but * would confirm that all the data is right.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 21]
```diff
@@ -808,6 +882,33 @@ private void assertTableLayout(Plan plan, String tableName, TupleDomain<Subfield
...
+    private void assertBucketFilter(Plan plan, String tableName, int readBucketCount, Set<Integer> bucketsToKeep)
+    {
+        TableScanNode tableScan = searchFrom(plan.getRoot())
+                .where(node -> isTableScanNode(node, tableName))
+                .findOnlyElement();
+
+        assertTrue(tableScan.getTable().getLayout().isPresent());
+        HiveTableLayoutHandle layoutHandle = (HiveTableLayoutHandle) tableScan.getTable().getLayout().get();
+
+        assertEquals(layoutHandle.getBucketHandle().get().getReadBucketCount(), readBucketCount);
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: @gparai add assertTrue(layoutHandle. getBucketHandle().isPresent()); and assertTrue(layoutHandle.getBucketFilter().isPresent());

Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 20 above)
Comment (gparai) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Yes, you are correct! I have modified the tests.

**Comment by mbasmanova** [comment]
@gparai Thank you for the contribution.
